ksqldb-engine/src/main/java/io/confluent/ksql/serde/SerdeFeaturesFactory.java

```diff
@@ -150,6 +161,20 @@ private static KeyFormat sanitizeKeyFormatForMultipleColumns(
     return KeyFormat.nonWindowed(FormatInfo.of(JsonFormat.NAME), SerdeFeatures.of());
   }
 
+  /**
+   * Check whether given sql type set needs a compound key support. It returns true
+   * when one of the conditions is met:
+   * 1. given sql types have more than one entry
+   * 2. it is a singleton non-primitive sql type
+   *
+   * @param sqlTypes the sql types to be tested
+   * @return true if given sql types do not need a compound key type support.
+   */
+  private static boolean needCompoundKeySupport(List<SqlType> sqlTypes) {
+    return sqlTypes.size() > 1 || (sqlTypes.size() == 1 &&
+        !(sqlTypes.get(0) instanceof SqlPrimitiveType));
```

Comment:
It's also not enough to check for primitive vs non-primitive types. This will work for the delimited format since the delimited format supports exactly primitive types and no non-primitive types, but this will not work for the Kafka format, which not only doesn't support non-primitive types, but also does not support some primitive types such as DECIMAL. We will need more custom logic to encode which types are supported by which formats. We should consider adding this into the Format interface, rather than encoding all the logic into this specific class.